Modules, Exports, and Imports.md

```diff
@@ -0,0 +1,301 @@
+
+Importing union types exposed follows the exact same syntax. For example, `import MyModule exposing (Action)` will
+import only the type, while `import MyModule exposing (Action(..))` will import any exported tags as well. You can
+also use exported tags qualified, like `MyModule.Increment`.
+
+## Opaque Types
+
+An opaque type is a union type where the type is exported but the tag(s) are not. Someone outside the module can see
+that the type exists, and pass it around, store it in records. They only thing they *can't* do is look inside the type,
+even if they know what the tags are named. Hence, it's opaque.
+
+Opaque types are Elm's way of enforcing information hiding. They allow a package author to define an interface of
+functions to create, update, and read useful values out of the type. The implementation can change completely, but as
+long as all functions on the type are updated to match, it's still considered a minor change. This gives package writers
+flexibility when writing their libraries. It also lets them rely on invariants, assured that the client hasn't meddled
+with values of the type in unexpected ways.
+
+Opaque types are less useful in applications. If you're typing to simply pass information around, exporting record type
+aliases is fine. If it makes sense to also define operations on these models, an opaque type might be a better fit.
+
+If your union type contains many tags, you can export functions that wrap them. You can be selective about which ones
+you export. Sometimes it can be tedious, but it does make a difference.
+
+```elm
+module Road (Action, addCar, redLight) where
+
+import Car exposing (Car)
+
+type Light =
+  Red | Yellow | Green
+
+type Action =
+  AddCar Car | Light Light | SomethingPrivate
+
+addCar : Car -> Action
+addCar =
+  AddCar
+
+redLight : Action
+redLight =
+  Light Red
+```
+
+You cannot make an opaque type out of a type alias; those are either exported or not, just like values. But you can
+create a union type to hide it. (This is more common when to opaque type represents a model rather than an action.)
+
+```elm
+module Person (Person, age) where
+
+type Person =
+  P { name : String, age : Int }
+
+age : Person -> Int
+age (P {age}) =
+  age
+```
+
+First, we define the union type `Person` with one tag, `P`, which carries a record. (You can put the record definition
+directly in the union type, or define an unexported alias.) Then we can export the `age` function which accesses the
+record in a way not possible outside of this module. The definition uses two nifty language features. First, it pattern
+matches on the `P` tag in the argument list. This is permitted when there is only one tag, because otherwise it's an
+incomplete pattern match. Next, `{age}` destructures the record, assigning the local constant `age` to the value of the
+record's `age` field. It's a much more concise way of writing this:
+
+```elm
+age person =
+  case person of
+    P record ->
+      record.age
+```
+
```

Comment:
I thought about this some more, and I'm not really sure it's a practice to encourage. I've only used it [once](http://package.elm-lang.org/packages/mgold/elm-turtle-graphics/1.0.2/Turtle) when I specifically wanted "for kids" and "for adults" modules. Otherwise I haven't really run into this. For example, there's enough going on in Random.PCG that you could arguably split off an internal module to make the main module read better. But I didn't and I think it's fine.